core/src/main/scala/kafka/utils/ShutdownableThread.scala

```diff
@@ -63,6 +65,8 @@ abstract class ShutdownableThread(val name: String, val isInterruptible: Boolean
         doWork()
       }
     } catch{
+      case fatalExitError: FatalExitError =>
+        fatalExitError.shutdownSystem()
```

Comment:
That is great! Not only you detected the issue but also you already have a solution. I have started thinking that perhaps we should add a unit test to reproduce this deadlock case. 

I guess this is a different issue form what we observed in KAFKA-3924. There a KafkaRequestHandler could not exit because it was waiting for partitionMapLock held by the thread that calls System.exit. But with the patch there should be no lock held when System.exit is invoked.

@resetius do you mind sharing the stack trace so we can study it in more detail?